source/opt/ccp_pass.cpp

```diff
@@ -0,0 +1,263 @@
+      return SSAPropagator::kNotInteresting;
+    }
+  }
+
+  // If there are no incoming executable edges, the meet ID will still be 0. In
+  // that case, return not interesting to evaluate the Phi node again.
+  if (meet_val_id == 0) {
+    return SSAPropagator::kNotInteresting;
+  }
+
+  // All the operands have the same constant value represented by |meet_val_id|.
+  // Set the Phi's result to that value and declare it interesting.
+  values_[phi->result_id()] = meet_val_id;
+  return SSAPropagator::kInteresting;
+}
+
+SSAPropagator::PropStatus CCPPass::VisitAssignment(ir::Instruction* instr) {
+  assert(instr->result_id() != 0 &&
+         "Expecting an instruction that produces a result");
+
+  // Instructions with no input operands will always result in a varying value.
+  if (instr->NumInOperands() == 0) {
+    return SSAPropagator::kVarying;
+  }
+
+  // If this is a copy operation, and the RHS is a known constant, assign its
+  // value to the LHS.
+  if (instr->opcode() == SpvOpCopyObject) {
+    uint32_t rhs_id = instr->GetSingleWordInOperand(0);
+    auto it = values_.find(rhs_id);
+    if (it != values_.end()) {
+      values_[instr->result_id()] = it->second;
+      return SSAPropagator::kInteresting;
+    }
+    return SSAPropagator::kNotInteresting;
+  }
+
+  // Otherwise, see if the RHS of the assignment folds into a constant value.
+  std::vector<uint32_t> cst_val_ids;
+  for (uint32_t i = 0; i < instr->NumInOperands(); i++) {
+    uint32_t op_id = instr->GetSingleWordInOperand(i);
+    auto it = values_.find(op_id);
+    if (it != values_.end()) {
+      cst_val_ids.push_back(it->second);
+    } else {
+      break;
+    }
+  }
+
+  // If we did not find a constant value for every operand in the instruction,
+  // do not bother folding it.  Indicate that this instruction does not produce
+  // an interesting value for now.  TODO(dnovillo): Some instructions will never
+  // produce an interesting constant value (e.g., most function calls).  Return
+  // kVarying for those.
+  auto constants = const_mgr_->GetConstantsFromIds(cst_val_ids);
+  if (constants.size() == 0) {
+    return SSAPropagator::kNotInteresting;
+  }
+
+  // Otherwise, fold the instruction with all the operands to produce a new
+  // constant.
+  uint32_t result_val = FoldScalars(instr->opcode(), constants);
+  const analysis::Constant* result_const =
+      const_mgr_->GetConstant(const_mgr_->GetType(instr), {result_val});
+  ir::Instruction* const_decl =
+      const_mgr_->GetDefiningInstruction(result_const);
+  values_[instr->result_id()] = const_decl->result_id();
+  return SSAPropagator::kInteresting;
+}
+
+SSAPropagator::PropStatus CCPPass::VisitBranch(ir::Instruction* instr,
+                                               ir::BasicBlock** dest_bb) const {
+  assert(instr->IsBranch() && "Expected a branch instruction.");
+  uint32_t dest_label = 0;
+
+  if (instr->opcode() == SpvOpBranch) {
```

Comment:
It is only three cases, but would this be cleaner as a switch?  Also, we could code this more defensively with a default case that assert for unknown branch instructions.  If a new branch is added with `IsBranch` changing, it will let people know they have to add a case.

Reply:
Well, there's not much to switch on.   The branches are protected against unknown branches in VisitBranch and the third case is not about specific opcodes.